### ISSUE TITLE
Agent reports if there is insufficient disk space during upgrade, cleans up, and prevents upgrade

### ISSUE THREAD

**Issue by ycombinator** [elastic/elastic-agent#5235]
When Agent tries to perform an upgrade, the local disk may not have sufficient space to a) download the new Agent artifact or b) to extract the downloaded Agent artifact.  When this happens, Agent should:
* reports this situation to the user (by reporting it to Fleet for Fleet-managed Agents), 
* cleans up any partially downloaded or extracted files, and
* prevents the upgrade from proceeding further

Acceptance critieria includes an E2E test proving the above steps are being taken.

**Comment by nimarezainia** [comment]
@cmacknz Is this complicated to achieve? I think everything is there we just need the message sent when this situation arises.
I believe the download clean up is also happening today.

cc: @ebeahan

**Comment by cmacknz** [comment]
It should not be complicated no. We will need to make sure we have a cross platform way to detect disk space on the correct drive (and permissions to do so) which will take some experimenting, but the upgrade status reporting to mark the upgrade as failed already exists.

**Comment by michalpristas** [comment]
my take is that we should still treat runtime failures as primary source of truth and doing some pre-emptive checks (optionally in one take or 2 PRs) based on disk size and content-length.
there are few issues with pre-checks. first free disk space can change, we can have process poluting disk space that would make upgrade fail. we can check content-length if repository reports it and fail if we don't have sufficient space for download, but we actually don't know what is the size of archive + extracted agent upfront. 
we need to figure out on which volume target path is located and support detection of free space across different OSes and filesystems (or use some 3rd party library). Statfs used in [go library](https://pkg.go.dev/golang.org/x/sys/unix) is using `statfs` syscall that was deprecated by LSB

**Comment by kaanyalti** [comment]
> my take is that we should still treat runtime failures as primary source of truth and doing some pre-emptive checks (optionally in one take or 2 PRs) based on disk size and content-length. there are few issues with pre-checks. first free disk space can change, we can have process poluting disk space that would make upgrade fail. we can check content-length if repository reports it and fail if we don't have sufficient space for download, but we actually don't know what is the size of archive + extracted agent upfront. we need to figure out on which volume target path is located and support detection of free space across different OSes and filesystems (or use some 3rd party library). Statfs used in [go library](https://pkg.go.dev/golang.org/x/sys/unix) is using `statfs` syscall that was deprecated by LSB

I agree with @Michal’s comment and wanted to share some thoughts and clarification questions.

**There are a few challenges with pre-checks:**
	•	Disk space can change at any moment, so even if we check it beforehand, a concurrent process could consume space before or during the upgrade, still resulting in failure.
	•	Content-Length may not always be available. The [downloader code](https://github.com/elastic/elastic-agent/blob/main/internal/pkg/agent/application/upgrade/artifact/download/http/downloader.go#L205-L210) already handles this case by falling back to runtime errors.
	•	We don’t know the final size on disk after download and extraction. So even with a known Content-Length, we’d still be guessing unless we can accurately estimate.

**Disk space calculation across OSes**

On Unix systems, I explored using `Statfs` [golang.org/x/sys/unix](https://pkg.go.dev/syscall). While it does indicate which volume a path is on, it uses the statfs syscall, which is deprecated by LSB https://cs.opensource.google/go/go/+/refs/tags/go1.24.5:src/syscall/zsyscall_linux_amd64.go;l=1350

For a basic test, I wrote this script: https://gist.github.com/kaanyalti/82877273e6b76639a120a4f0fa1e3b96. The results were insightful:

```
=== Report: Disk full write comparison (reserved=0%) ===
Block size: 4096 bytes
User                 ExpectedBlocks  BlocksWritten  DiffBlocks  BytesExpected  BytesWritten  DiffBytes
-------------------- -------------- ------------- ---------- -------------- ------------- ----------
root                          22954         22442        512       94019584      91922432    2097152
unprivileged-user             22442         22441          1       91922432      91918336       4096

=== Report: Disk full write comparison (reserved=10%) ===
Block size: 4096 bytes
User                 ExpectedBlocks  BlocksWritten  DiffBlocks  BytesExpected  BytesWritten  DiffBytes
-------------------- -------------- ------------- ---------- -------------- ------------- ----------
root                          22954         22441        513       94019584      91918336    2101248
unprivileged-user             19882         19881          1       81436672      81432576       4096

=== Final Comparison Report ===
Test Case           User                 ExpectedBlocks  BlocksWritten  DiffBlocks  BytesExpected  BytesWritten  DiffBytes
------------------- -------------------- -------------- ------------- ---------- -------------- ------------- ----------
reserved=0%         root                          22954         22442        512       94019584      91922432    2097152
reserved=0%         unprivileged-user             22442         22441          1       91922432      91918336       4096
reserved=10%        root                          22954         22441        513       94019584      91918336    2101248
reserved=10%        unprivileged-user             19882         19881          1       81436672      81432576       4096
``` 
When no space is reserved for root (reserved=0%), both root and non-root users can access Bavail. When space is reserved (reserved=10%), root can access more than Bavail, but still less than Bfree. So, it’s unclear how we can accurately determine available space for the root user, other than by making a conservative assumption.

I haven’t tested this on Windows yet, but the API there seems more straightforward, it returns the available space for the calling user directly.

**Clarification question:**

Looking at the [downloader logic](https://github.com/elastic/elastic-agent/blob/main/internal/pkg/agent/application/upgrade/artifact/download/http/downloader.go#L216-L222), the agent already appears to handle disk space issues at runtime by checking errors returned from `io.Copy`. I haven’t dug into the extraction step yet, but if it’s handled similarly, is the goal of this issue to improve the user-facing error message rather than introduce full-blown pre-checks?

In my opinion, as @michalpristas also suggested, the agent should rely on runtime failures.

cc: @ycombinator @cmacknz @nimarezainia @ebeahan

**Comment by cmacknz** [comment]
> reports this situation to the user (by reporting it to Fleet for Fleet-managed Agents),
> 
> cleans up any partially downloaded or extracted files, and
> 
> prevents the upgrade from proceeding further

The steps in the description don't require a pre-check and would still be satisfied if we attempted the upgrade and set the upgrade state to failed with a clear and standardized "out of disk space" message and tidied up after the failure as we should already be doing.

We already have an integration that can report disk space across OSes (system), if we wanted to go a bit further we could turn it on as part of self-monitoring and show a disk full indicator in agent details along side agent cpu + memory for the volume agent is installed on. This would be a separate scope of work.

**Comment by cmacknz** [comment]
There are three points where agent can fail due to disk space errors I think, not just two:

1. When downloading the compressed archive
2. When extracting the compressed archive
3. When starting the agent for the first time at which point it will use additional disk by writing to its log files and local state.

The third one would probably result in an upgrade rollback and would be harder to get a standardized error from.

**Comment by michalpristas** [comment]
the third one is also a use case not specific to upgrade, we could think about reporting out of disk as an error state to fleet for visibility (degraded + message)

**Comment by cmacknz** [comment]
I think the first two cases will be by far the most common where you need a few hundred MB of space for the download and ~1.5 GB to extract on Linux. The third failure is what you'd hit if you had a couple of megabytes or less of free disk space.

What I'd suggest is focusing on the first two in https://github.com/elastic/elastic-agent/issues/5235#issuecomment-3079768395 and reporting via upgrade details on failure and then we fork the 3rd case out into a separate problem. That this third case isn't specific to upgrade is a great point.

**Comment by kaanyalti** [comment]
I ran a quick test to see how the agent behaves when it tries to download the agent artifacts while upgrading. Reading the description of this issue, I was under the impression that the agent is not surfacing any disk space error in any way when it runs into space problems. This contradicts what I see in the code, specifically in the http downloader code. As I had suspected, the agent indeed reports insufficient disk space. However, after reporting "no space left on device", the agent keeps retrying and eventually the context times out, at which point the disk space error gets masked by the context deadline error. As a result the disk space error does not show up on fleet ui. It is possible to see the insufficient disk space error in the status report though as seen below

```
┌─ fleet
│  └─ status: (HEALTHY) Connected
├─ elastic-agent
│  └─ status: (HEALTHY) Running
└─ upgrade_details
   ├─ target_version: 9.0.4
   ├─ state: UPG_FAILED
   ├─ action_id: dad9e942-1812-4fe3-8c57-a29e87a094a8
   └─ metadata
      ├─ failed_state: UPG_DOWNLOADING
      ├─ error_msg: failed download of agent binary: context deadline exceeded
      ├─ retry_until: -19m6.933408225s
      └─ retry_error_msg: unable to download package: package '/opt/Elastic/Agent/data/elastic-agent-8.10.0-SNAPSHOT-58886e/downloads/elastic-agent-9.0.4-linux-arm64.tar.gz' not found: open /opt/Elastic/Agent/data/el
astic-agent-8.10.0-SNAPSHOT-58886e/downloads/elastic-agent-9.0.4-linux-arm64.tar.gz: no such file or directory
         copying fetched package failed: write /opt/Elastic/Agent/data/elastic-agent-8.10.0-SNAPSHOT-58886e/downloads/elastic-agent-9.0.4-linux-arm64.tar.gz: no space left on device
```

I don't see a reason why the agent should keep trying if there isn't enough space on disk. Available disk space can change as the agent is working on the upgrade download; however, aborting the upgrade with a failed status would be simpler. The other option would be to let the agent retry downloading and remember the disk space error so that it does not get covered up by the context deadline error.

@cmacknz @michalpristas

**Comment by nimarezainia** [comment]
I think the conversation is going down this path anyway, but to reiterate, I don't think we need a pre-check before proceeding. The desired user outcome can be achieved if we:

- proceed with the download, in the event of a failure (not enough space) add a log and make this error visible in Fleet.
- our new automatic upgrade facility will attempt the upgrade again, if user wants everything to be automatic
- the user can also select to upgrade again after the disk space issue has been mitigated

**Comment by michalpristas** [comment]
do we have some watcher in mind that would be triggered when insufficient disk error is raised that would notify fleet/agent about disk space cleanup? out of scope of this issue of course. 
this would nicely complement automatic retry

**Comment by nimarezainia** [comment]
@michalpristas regarding the *watcher* , I was referring to [this](https://github.com/elastic/ingest-dev/issues/2878) which was merged for 9.1

**Comment by kaanyalti** [comment]
In my comment above https://github.com/elastic/elastic-agent/issues/5235#issuecomment-3111589765 I'm not sure if I was able to convey what I actually wanted to clarify. Currently the agent is already fulfilling the requirements of the issue when downloading the new agent artifact. If the agent fails to download the artifact due to insufficient disk space it properly reports this in the upgrade details, which shows up in fleet ui. The problem is the insufficient disk space message gets replaced by context deadline exceeded error. This is happening because the agent retries downloading until the retry times out. The default timeout is 2 hours. So the customer should see upgrade failure with insufficient disk message in fleet ui for at least 2 hours, after which it gets replaced with context deadline message.

I can change the behaviour so that:
1. The agent exposes the insufficient disk error indefinitely even after download retry times out
2. The agent exposes any underlying download error indefinitely even after download retry times out
3. The agent will stop retrying downloads immediately and mark the upgrade as failed if it runs into insufficient disk error.

To clarify, the retry I'm talking about happens in this function https://github.com/elastic/elastic-agent/blob/main/internal/pkg/agent/application/upgrade/step_download.go#L220

@nimarezainia @michalpristas @cmacknz 


Here's a snapshot of insufficient disk message in fleet ui

<img width="1738" height="706" alt="Image" src="https://github.com/user-attachments/assets/6da3b32c-e726-4c38-99a3-93e219e2a133" />

and the corresponding status output
```
┌─ fleet
│  └─ status: (HEALTHY) Connected
├─ elastic-agent
│  └─ status: (UPGRADING) Upgrading to version 9.0.4
└─ upgrade_details
   ├─ target_version: 9.0.4
   ├─ state: UPG_FAILED
   ├─ action_id: 10bfe9ee-d365-4971-8090-a0246867baec
   └─ metadata
      ├─ failed_state: UPG_DOWNLOADING
      ├─ error_msg: write /opt/Elastic/Agent/data/elastic-agent-8.10.0-SNAPSHOT-58886e/downloads/elastic-agent-9.0.4-linux-arm64.tar.gz: no space left on device
      ├─ retry_until: 11.93017993s
      └─ retry_error_msg: unable to download package: package '/opt/Elastic/Agent/data/elastic-agent-8.10.0-SNAPSHOT-58886e/downloads/elastic-agent-9.0.4-linux-arm64.tar.gz' not found: open /opt/Elastic/Agent/data/elastic-agent-8.10.0-SNAPS
HOT-58886e/downloads/elastic-agent-9.0.4-linux-arm64.tar.gz: no such file or directory
         copying fetched package failed: write /opt/Elastic/Agent/data/elastic-agent-8.10.0-SNAPSHOT-58886e/downloads/elastic-agent-9.0.4-linux-arm64.tar.gz: no space left on device
```

And here's the screenshot of fleet ui after the download retries timeout

<img width="1589" height="650" alt="Image" src="https://github.com/user-attachments/assets/e3e0993f-d3e3-487c-8f61-269d64ad38d5" />

and the corresponding status output 

```
┌─ fleet
│  └─ status: (HEALTHY) Connected
├─ elastic-agent
│  └─ status: (HEALTHY) Running
└─ upgrade_details
   ├─ target_version: 9.0.4
   ├─ state: UPG_FAILED
   ├─ action_id: 10bfe9ee-d365-4971-8090-a0246867baec
   └─ metadata
      ├─ failed_state: UPG_DOWNLOADING
      ├─ error_msg: failed download of agent binary: context deadline exceeded
      ├─ retry_until: -6m42.078298707s
      └─ retry_error_msg: unable to download package: package '/opt/Elastic/Agent/data/elastic-agent-8.10.0-SNAPSHOT-58886e/downloads/elastic-agent-9.0.4-linux-arm64.tar.gz' not found: open /opt/Elastic/Agent/data/elastic-agent-8.10.0-SNAPS
HOT-58886e/downloads/elastic-agent-9.0.4-linux-arm64.tar.gz: no such file or directory
         copying fetched package failed: write /opt/Elastic/Agent/data/elastic-agent-8.10.0-SNAPSHOT-58886e/downloads/elastic-agent-9.0.4-linux-arm64.tar.gz: no space left on device
```

**Comment by nimarezainia** [comment]
tbh I think the user requirements are being addressed. I don't know if we should be removing the retries.

**Comment by michalpristas** [comment]
i think we can easily address error masking by making disk size failure non retriable and fail immediatelly. 
then error visible in fleet would be the proper one

**Comment by kaanyalti** [comment]
> tbh I think the user requirements are being addressed. I don't know if we should be removing the retries.

I want to clarify that I was suggesting breaking out of retries only if there is not enough space on disk, otherwise I do agree that we should keep retries.

I had a quick discussion with @pkoutsovasilis and came up with the following approach for this issue:

1. The download retry should not be masking the underlying error. The user should only see a context timeout error if the download times out due to slow network. Otherwise the user should be able to see the underlying error.
  1a. Agent is already cleaning up after failed downloads https://github.com/elastic/elastic-agent/blob/main/internal/pkg/agent/application/upgrade/upgrade.go#L244
2. There are several steps after downloading the artifact where the agent is not cleaning up the downloaded content after errors. This link shows them all https://github.com/elastic/elastic-agent/blob/main/internal/pkg/agent/application/upgrade/upgrade.go#L253-L315
  2a. Agent should be cleaning up the downloaded files if there are errors in the code block shown in the link
  2b. After the linked code block, the agent is already cleaning up the downloaded files if there are any errors.

In addition to this I want to show the result of the unpacking test that I did. I filled up the disk and left enough space for the agent to download the new agent artifact, and then triggered an upgrade through fleet. 

<img width="1563" height="624" alt="Image" src="https://github.com/user-attachments/assets/55a18f1f-a208-4c8d-98e1-7f5607848267" />

and the corresponding status output

```
┌─ fleet
│  └─ status: (HEALTHY) Connected
├─ elastic-agent
│  └─ status: (HEALTHY) Running
└─ upgrade_details
   ├─ target_version: 9.0.4
   ├─ state: UPG_FAILED
   ├─ action_id: 3908b8f5-7125-4bde-82f4-f2d313ef9398
   └─ metadata
      ├─ failed_state: UPG_EXTRACTING
      └─ error_msg: TarInstaller: error writing to /opt/Elastic/Agent/data/elastic-agent-9.0.4-d49717/elastic-agent: write /opt/Elastic/Agent/data/elastic-agent-9.0.4-d49717/elastic-agent: no space left on device
```

Contrary to the download errors, the errors from unpacking persist and don't get replaced by any other error.

@cmacknz @nimarezainia @michalpristas

**Comment by cmacknz** [comment]
Agree with the above, the only note would be we should shorten the error in `TarInstaller: error writing to /opt/Elastic/Agent/data/elastic-agent-9.0.4-d49717/elastic-agent: write /opt/Elastic/Agent/data/elastic-agent-9.0.4-d49717/elastic-agent: no space left on device` to just "no space left on device" or similar if we can, and use the same shortened message consistently.

**Comment by kaanyalti** [comment]
> Agree with the above, the only note would be we should shorten the error in `TarInstaller: error writing to /opt/Elastic/Agent/data/elastic-agent-9.0.4-d49717/elastic-agent: write /opt/Elastic/Agent/data/elastic-agent-9.0.4-d49717/elastic-agent: no space left on device` to just "no space left on device" or similar if we can, and use the same shortened message consistently.

This should be fairly simple by checking for disk space related errno when there is an error.

**Comment by kaanyalti** [comment]
Opened the following PRs
- #9122  
- #9322 
- #9366 
- #9349  
- #9386 
- #9392

**Comment by ycombinator** [comment]
> Opened the following PRs

Are any / all of these ready to review or are you still working on them? Asking because I've seen you pushing up several commits so it's not clear if you're still working on these or they are ready to review.

**Comment by kaanyalti** [comment]
> > Opened the following PRs
> 
> Are any / all of these ready to review or are you still working on them? Asking because I've seen you pushing up several commits so it's not clear if you're still working on these or they are ready to review.

They are all ready to review

**Comment by kaanyalti** [comment]
I would like to provide an overview of what each PR does and the decisions I took to provide the reviewers with better context

To help myself make sense of what's happening in the upgrade process and to divide my work into meaningful chunks I mentally split the Upgrade function into several logical steps:
- Download
- Unpack
- Replace old agent with new agent
	- Directory migrations
	- Symlink switch
- Watch new agent

There is one PR per each of these logical steps to make sure that the errors in these steps get propagated up the call stack.

- https://github.com/elastic/elastic-agent/pull/9122
	- In addition to updating error propagation, this PR also updates the downloader to stop retrying if it runs into a insufficient disk space error.
- https://github.com/elastic/elastic-agent/pull/9322
- https://github.com/elastic/elastic-agent/pull/9349
- https://github.com/elastic/elastic-agent/pull/9366

In addition, there are these two PRs:
- https://github.com/elastic/elastic-agent/pull/9386 adds a cleanup step to the upgrade function so that the agent removes any downloaded artifacts or extracted agent files if there are errors along the way. There is a gap in cleanup in the current implementation and this PR addresses that.
- https://github.com/elastic/elastic-agent/pull/9392 adds an additional error check in the coordinator to look for insufficient disk space errors in order to be able to provide the user with a cleaner error message.

All the PRs listed above can be reviewed independently. 

The disk space error type and stdlib abstractions are introduced in multiple PRs so that each PR can be reviewed independently. As soon as one of the PRs get merged, rebasing the others on main will get rid of this repetition.

The disk space error files are repeated in
- https://github.com/elastic/elastic-agent/pull/9322
- https://github.com/elastic/elastic-agent/pull/9122
- https://github.com/elastic/elastic-agent/pull/9366
- https://github.com/elastic/elastic-agent/pull/9349
- https://github.com/elastic/elastic-agent/pull/9392

The stdlib wrappers and the mock util are repeated in
- https://github.com/elastic/elastic-agent/pull/9322
- https://github.com/elastic/elastic-agent/pull/9122
- https://github.com/elastic/elastic-agent/pull/9366
- https://github.com/elastic/elastic-agent/pull/9349


Tagging all the reviewers here so that this comment is more visible:
@michalpristas @blakerouse @ycombinator @pkoutsovasilis @michel-laterman

**Comment by kaanyalti** [comment]
After discussing the with the team, I converted all the PRs except for #9122 into draft PRs. I'm not going to try to get all of them reviewed all in parallel, which caused quite a bit of confusion. Once #9122 is merged I will put the rest of the PRs in review one by one.

@ycombinator @michalpristas @pkoutsovasilis @blakerouse @michel-laterman
